### ISSUE TITLE
Intermittent false negatives when using "is T v" where T is tuple

### ISSUE THREAD

**Issue by kimbirkelund** [dotnet/roslyn#26387]
**Version Used**: netcoreapp2.0 using dotnet CLI 15.6.0-preview-20180109-01, also observed on Ubuntu with .NET Core 2.0 and Windows with .NET Core 2.0 and Visual Studio 15.3.3 and donet CLI 2.0.0.

**Steps to Reproduce**:

Using this .csproj: 

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.7.0" />
    <PackageReference Include="xunit" Version="2.3.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.3.1" />
  </ItemGroup>

</Project>
```

and this .cs file:

```
using System;
using Xunit;

namespace ConsoleApp1
{
    public class TestHelper
    {
        [Fact]
        public void TestHelperNeverReturnsNull()
        {
            IsValueTypeT(new Result<(Cls, IInt)>((new Cls(), new Int())));
        }

        [Fact]
        public void TestHelperNeverReturnsNull_Continously()
        {
            Assert.Throws<NotPossibleException>(
                new Action(() =>
                {
                    while (true)
                        TestHelperNeverReturnsNull();
                }));
        }

        public static void IsValueTypeT<T>(Result<T> result)
        {
            if (!(result.Value is T v))
                throw new NotPossibleException();
        }
    }

    public class Result<T>
    {
        public T Value { get; }

        public Result(T value)
        {
            Value = value;
        }
    }

    public class Cls { }
    public interface IInt { }

    public class Int : IInt { }

    public class NotPossibleException : Exception { }
}
```

Running `dotnet test --configuration Release`. Gives output

```
Build started, please wait...                                                                                                                                                                     
Program.cs(25,28): warning xUnit1013: Public method 'IsValueTypeT' on test class 'TestHelper' should be marked as a Theory. [C:\Users\kim\source\repos\ConsoleApp1\ConsoleApp1\ConsoleApp1.csproj]
Build completed.                                                                                                                                                                                  
                                                                                                                                                                                                  
Test run for C:\Users\kim\source\repos\ConsoleApp1\ConsoleApp1\bin\Release\netcoreapp2.0\ConsoleApp1.dll(.NETCoreApp,Version=v2.0)                                                                
Microsoft (R) Test Execution Command Line Tool Version 15.6.0-preview-20180109-01                                                                                                                 
Copyright (c) Microsoft Corporation.  All rights reserved.                                                                                                                                        
                                                                                                                                                                                                  
Starting test execution, please wait...                                                                                                                                                           
[xUnit.net 00:00:00.3542121]   Discovering: ConsoleApp1                                                                                                                                           
[xUnit.net 00:00:00.4069510]   Discovered:  ConsoleApp1                                                                                                                                           
[xUnit.net 00:00:00.4122708]   Starting:    ConsoleApp1                                                                                                                                           
[xUnit.net 00:00:00.5435184]     ConsoleApp1.TestHelper.TestHelperNeverReturnsNull [FAIL]                                                                                                         
[xUnit.net 00:00:00.5446641]       ConsoleApp1.NotPossibleException : Exception of type 'ConsoleApp1.NotPossibleException' was thrown.                                                            
[xUnit.net 00:00:00.5458188]       Stack Trace:                                                                                                                                                   
[xUnit.net 00:00:00.5471366]         C:\Users\kim\source\repos\ConsoleApp1\ConsoleApp1\Program.cs(28,0): at ConsoleApp1.TestHelper.IsValueTypeT[T](Result`1 result)                               
[xUnit.net 00:00:00.5605291]   Finished:    ConsoleApp1                                                                                                                                           
Failed   ConsoleApp1.TestHelper.TestHelperNeverReturnsNull                                                                                                                                        
Error Message:                                                                                                                                                                                    
 ConsoleApp1.NotPossibleException : Exception of type 'ConsoleApp1.NotPossibleException' was thrown.                                                                                              
Stack Trace:                                                                                                                                                                                      
   at ConsoleApp1.TestHelper.IsValueTypeT[T](Result`1 result) in C:\Users\kim\source\repos\ConsoleApp1\ConsoleApp1\Program.cs:line 28                                                             
                                                                                                                                                                                                  
Total tests: 2. Passed: 1. Failed: 1. Skipped: 0.                                                                                                                                                 
Test Run Failed.                                                                                                                                                                                  
Test execution time: 1,2055 Seconds                                                                                                                                                               

```

This code creates a tuple in a wrapper the type of which is then tested in `IsValueTypeT` using a type parameter. This should never fail - or fail (at compile time) if it wasn't allowed. 

I've also seen this behavior outside of an xunit test, and I've reproduced it in Visual Studio 2017 15.6.6 in Debug configuration. In CLI I couldn't reproduce in Debug configuration.

**Expected Behavior**:

TestHelperNeverReturnsNull should pass and TestHelperNeverReturnsNull_Continously should never finish.

**Actual Behavior**:

TestHelperNeverReturnsNull  passes most of the time and TestHelperNeverReturnsNull_Continously completes at some point (indicating that the NotPossibleException was thrown).

**Comment by svick** [comment]
Simplified reproduction:

.csproj:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>
</Project>
```

Program.cs:

```c#
using System;

class TestHelper
{
    static void Main() => IsValueTypeT(new S());

    static void IsValueTypeT<T>(T result)
    {
        if (!(result is T v))
            throw new Exception();
    }
}

struct S { }
```

Executing this with `dotnet run` produces:

```
Unhandled Exception: System.Exception: Exception of type 'System.Exception' was thrown.
   at TestHelper.IsValueTypeT[T](T result) in C:\code\tmp\hwapp\Program.cs:line 11
   at TestHelper.Main() in C:\code\tmp\hwapp\Program.cs:line 5
```

The IL for `IsValueTypeT` in Release mode is:

```
.method private hidebysig static void  IsValueTypeT<T>(!!T result) cil managed
{
  // Code size       19 (0x13)
  .maxstack  2
  .locals init (!!T V_0)
  IL_0000:  ldarg.0
  IL_0001:  ldloca.s   V_0
  IL_0003:  initobj    !!T
  IL_0009:  ldloc.0
  IL_000a:  bne.un.s   IL_0012
  IL_000c:  newobj     instance void [System.Runtime]System.Exception::.ctor()
  IL_0011:  throw
  IL_0012:  ret
} // end of method TestHelper::IsValueTypeT
```

This doesn't look right: you can't compare two `T`s using `bne.un`. And even if you could, comparing with `default` doesn't make sense here.

And peverify agrees this is not correct code, its output (after changing target framework to `net471`) is:

```
Microsoft (R) .NET Framework PE Verifier.  Version  4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

[IL]: Error: [C:\code\tmp\hwapp\bin\Release\net471\hwapp.exe : TestHelper::IsValueTypeT[T]][offset 0x0000000A][found (unboxed) 'T'] Non-compatible types on the stack.
1 Error(s) Verifying hwapp.exe
```

**Comment by kimbirkelund** [comment]
Does this reproduce on every run? And if not do you have an explanation for why?

**Comment by svick** [comment]
@kimbirkelund Yes, it does reproduce for me on every run for my simplified repro.

**Comment by kimbirkelund** [comment]
Do you have an explanation for why mine doesn't? In the app where the bug originated it occured at an interval of 1 in 50 invocations or something like that - I was able to reproduce prior to trying to get detail of the issue in production. In any case it doesn't occur on every invocation, and when it occurs it doesn't keep occurring on all subsequent occurrence. 

I'm guessing that it runs long enough to be optimized by the JIT and then fails, only to be optimized at a later stage. But if your repro fails on the first invocation that would indicate that it either hits the optimize path earlier or that its another path.

**Comment by gafter** [comment]
This bug is already fixed on the `features/recursive-patterns` branch, but broken in `master`.

**Comment by gafter** [comment]
Fix in review in https://github.com/dotnet/roslyn/pull/26390
